#! /bin/sh /usr/share/dpatch/dpatch-run
## dynamic-hostname.dpatch by  <martin.pitt@ubuntu.com>
##
## DP: Add support for a new string type 'h' which behaves like 't' except that
## DP: '<hostname>' is changed to the current hostname.

@DPATCH@
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' isc-dhcp~/common/options.c isc-dhcp/common/options.c
--- isc-dhcp~/common/options.c	2009-07-23 20:02:09.000000000 +0100
+++ isc-dhcp/common/options.c	2010-11-01 17:24:04.000000000 +0000
@@ -1501,6 +1501,7 @@
 		switch (*p++) {
 		    case 'd':
 		    case 't':
+		    case 'h':
 			return 1;
 
 			/* These symbols are arbitrary, not fixed or
@@ -1626,6 +1627,7 @@
 		    case 'd': /* "Domain name" */
 		    case 'D': /* "rfc1035 formatted names" */
 		    case 't': /* "ASCII Text" */
+		    case 'h': /* "ASCII Text with <hostname> magic" */
 		    case 'X': /* "ASCII or Hex Conditional */
 		    case 'x': /* "Hex" */
 		    case 'A': /* Array of all that precedes. */
@@ -1750,6 +1752,7 @@
 			fmtbuf[l] = 't';
 			/* Fall Through ! */
 		      case 't':
+		      case 'h':
 			fmtbuf[l + 1] = 0;
 			numhunk = -2;
 			break;
@@ -1837,6 +1840,7 @@
 		for (j = 0; j < numelem; j++) {
 			switch (fmtbuf [j]) {
 			      case 't':
+			      case 'h':
 				/* endbuf-1 leaves room for NULL. */
 				k = pretty_text(&op, endbuf - 1, &dp,
 						data + len, emit_quotes);
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' isc-dhcp~/common/parse.c isc-dhcp/common/parse.c
--- isc-dhcp~/common/parse.c	2009-07-23 20:02:09.000000000 +0100
+++ isc-dhcp/common/parse.c	2010-11-01 17:24:16.000000000 +0000
@@ -34,6 +34,7 @@
 
 #include "dhcpd.h"
 #include <syslog.h>
+#include <limits.h>
 
 /* Enumerations can be specified in option formats, and are used for
    parsing, so we define the routines that manage them here. */
@@ -5043,6 +5044,23 @@
 	return 1;
 }
 
+int make_possible_hostname(val, len)
+	char** val;
+	unsigned* len;
+{
+	static char localhostname[HOST_NAME_MAX];
+
+	if (!strcmp(*val, "<hostname>")) {
+		if (gethostname (localhostname, HOST_NAME_MAX))
+			return 0;
+		localhostname [HOST_NAME_MAX-1] = 0;
+		*val = localhostname;
+		*len = strlen(localhostname);
+	}
+	return 1;
+}
+
+
 int parse_option_token (rv, cfile, fmt, expr, uniform, lookups)
 	struct expression **rv;
 	struct parse *cfile;
@@ -5144,6 +5162,7 @@
 		goto make_string;
 
 	      case 't': /* Text string... */
+	      case 'h': /* Text string with <hostname> magic... */
 		token = next_token (&val, &len, cfile);
 		if (token != STRING && !is_identifier (token)) {
 			if ((*fmt) [1] != 'o') {
@@ -5153,6 +5172,9 @@
 			}
 			return 0;
 		}
+		if (**fmt == 'h' && !make_possible_hostname (&val, &len))
+			parse_warn (cfile, "resolving <hostname> failed");
+
 	      make_string:
 		if (!make_const_data (&t, (const unsigned char *)val,
 				      len, 1, 1, MDL))
@@ -5364,6 +5386,7 @@
 				break;
 					
 			      case 't': /* Text string... */
+			      case 'h': /* Text string with <hostname> magic... */
 				token = peek_token (&val,
 						    &len, cfile);
 				if (token == SEMI && fmt[1] == 'o') {
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' isc-dhcp~/common/tables.c isc-dhcp/common/tables.c
--- isc-dhcp~/common/tables.c	2009-07-24 23:04:52.000000000 +0100
+++ isc-dhcp/common/tables.c	2010-11-01 17:24:04.000000000 +0000
@@ -59,6 +59,8 @@
    b - 8-bit signed integer
    B - 8-bit unsigned integer
    t - ASCII text
+   h - same as t, but if the ASCII string is "<hostname>", option is set
+       to the value returned by gethostname(2)
    T - Lease Time, 32-bit unsigned integer implying a number of seconds from
        some event.  The special all-ones value means 'infinite'.  May either
        be printed as a decimal, eg, "3600", or as this name, eg, "infinite".
@@ -108,7 +110,7 @@
 	{ "lpr-servers", "IA",			&dhcp_universe,   9, 1 },
 	{ "impress-servers", "IA",		&dhcp_universe,  10, 1 },
 	{ "resource-location-servers", "IA",	&dhcp_universe,  11, 1 },
-	{ "host-name", "t",			&dhcp_universe,  12, 1 },
+	{ "host-name", "h",			&dhcp_universe,  12, 1 },
 	{ "boot-size", "S",			&dhcp_universe,  13, 1 },
 	{ "merit-dump", "t",			&dhcp_universe,  14, 1 },
 	{ "domain-name", "t",			&dhcp_universe,  15, 1 },
